torchtune/utils/_device.py

```diff
@@ -5,9 +5,50 @@
 # LICENSE file in the root directory of this source tree.
 
 import os
-from typing import Optional
+from typing import Any, Optional, cast
 
 import torch
+if torch.xpu.is_available():
+    import intel_extension_for_pytorch
+
+class _DeviceHandle:
+    """
+    This is a simple abstraction for computing devices,
+    which enables custom backends that implement CUDA-like
+    semantics.
```

Comment:
Could you provide a few more details here on how this should be used? MY understanding is that this is used in place of device. A few examples would be helpful.

Reply:
This `device_handle`is designed to eliminate redundant code when implementing CUDA-like APIs for your customized device. For instance, consider the XPU. To empty the cache, a simplistic approach involves using conditional statements:

`torch.xpu.empty_cache() if device == 'xpu' else torch.cuda.empty_cache()`

This method, however, lacks efficiency and flexibility, particularly as the number of GPU devices increases in the future. Therefore, we adopt the `device_handle` strategy to integrate both CUDA APIs and the CUDA-like APIs of the customized device.

To utilize this strategy, you should first define your `device_handle` based on your device:

`device_handle = get_device_handle (device) # device could be ‘cuda’, ‘xpu’ or your customed device`

Then, the code for emptying the cache can be rewritten as:

`device_handle.empty_cache()`

In this setup, `device_handle` replaces direct calls to `torch.cuda` or `torch.xpu`, streamlining the process across different device types. Commonly used CUDA APIs, such as `torch.cuda.memory_allocated()` and `torch.cuda.synchronize()`, can be utilized in the same manner.